components/brave_wallet_ui/common/trezor/trezor-messages.ts

```diff
@@ -0,0 +1,192 @@
+  command: TrezorCommand.GetAccounts,
+  paths: TrezorAccountPath[]
+}
+export type UnlockCommand = CommandMessage & {
+  command: TrezorCommand.Unlock
+}
+export type UnlockResponse = CommandMessage & {
+  result: Boolean,
+  error?: Unsuccessful
+}
+export type TrezorAccount = {
+  publicKey: string
+  serializedPath: string,
+  fingerprint: number
+}
+export type TrezorError = {
+  error: string,
+  code: string
+}
+export type TrezorGetAccountsResponse = {
+  success: Boolean
+  payload: TrezorAccount[] | TrezorError
+}
+export type GetAccountsResponsePayload = CommandMessage & {
+  payload: TrezorGetAccountsResponse
+}
+export type TrezorFrameCommand = GetAccountsCommand | UnlockCommand
+export type TrezorFrameResponse = UnlockResponse | GetAccountsResponsePayload
+
+// Trezor library is loaded inside the chrome-untrusted webui page
+// and communication is going through posting messages between parent window
+// and frame window. This class handles low level messages transport to add,
+// remove callbacks and allows to process messages for childrens.
+abstract class MessagingTransport {
+  constructor () {
+    this.handlers = new Map<string, Function>()
+  }
+
+  handlers: Map<string, Function>
+
+  addCommandHandler = (id: string, listener: Function): Boolean => {
+    if (!this.handlers.size) {
+      this.addWindowMessageListener()
+      this.handlers.clear()
+    }
+    if (this.handlers.has(id)) {
+      return false
+    }
+    this.handlers.set(id, listener)
+    return true
+  }
+
+  removeCommandHandler = (id: string) => {
+    if (!this.handlers.has(id)) {
+      return false
+    }
+    this.handlers.delete(id)
+    if (!this.handlers.size) {
+      this.removeWindowMessageListener()
+    }
+    return true
+  }
+
+  abstract onMessageReceived (event: MessageEvent): unknown
+
+  private addWindowMessageListener = () => {
+    window.addEventListener('message', this.onMessageReceived)
+  }
+
+  private removeWindowMessageListener = () => {
+    window.removeEventListener('message', this.onMessageReceived)
+  }
+}
+
+// Handles sending messages to the Trezor library, creates untrusted iframe,
+// loads library and allows to send commans to the library and subscribe
+// for responses.
+export class TrezorBridgeTransport extends MessagingTransport {
+  constructor (bridgeFrameUrl: string) {
+    super()
+    this.bridgeFrameUrl = bridgeFrameUrl
+    // @ts-ignore
+    this.frameId = crypto.randomUUID()
+  }
+
+  frameId: string
+  bridgeFrameUrl: string
+
+  sendCommandToTrezorFrame = async (command: TrezorFrameCommand, listener: Function): Promise<Boolean> => {
+    let bridge = this.getBridge() as HTMLIFrameElement
+    if (!bridge) {
+      bridge = await this.createBridge() as HTMLIFrameElement
+    }
+    if (!bridge.contentWindow) {
+      return false
+    }
+    this.addCommandHandler(command.id, listener)
+
+    bridge.contentWindow.postMessage(command, this.bridgeFrameUrl)
+    return true
+  }
```

Comment:
Future improvement ideas (for follow-up and only if you want to do them, don't have to do now):
1. Generate the message ID here, instead of relying on every caller to generate a unique message ID. This function parameter would then be: `command: Omit<TrezorFrameCommand, 'id'>`
2. Make listener optional, or:
3. Instead of accepting a listener callback, add a promise resolve that the caller can await which will even further simplify calling this. Implementation would look roughly like:
```ts
  // T is response type, e.g. UnlockResponse. Resolves as `false` if transport errro
  sendCommandToTrezorFrame = async<T> (command: TrezorFrameCommand): Promise<T | false> => {
    return new Promise<T>(async (resolve) => {
      let bridge = this.getBridge()
      if (!bridge) {
        bridge = await this.createBridge()
      }
      if (!bridge.contentWindow) {
        Promise.resolve(false)
        return
      }
      this.addCommandHandler(command.id, resolve)
      bridge.contentWindow.postMessage(command, this.bridgeFrameUrl)
    })
  }
```

Reply:
done